### PR TITLE
fix: getConstructor name working with PoorlyConstructedErrors

### PR DESCRIPTION
**Pull request by lucasfcosta**
Hello everyone,

Do you guys remember about [this comment right here](https://github.com/chaijs/check-error/pull/4#issuecomment-223279797)? Turns out I was wrong.

I have forgotten about why I needed `new` there.
Actually we do needed the `new` keyword there because of poorly-constructed errors like the ones described in [this issue](https://github.com/chaijs/chai/issues/45).

I found this out when updating [this PR on Chai](https://github.com/chaijs/chai/pull/683).

The only fix which could be done is to check if we've got an empty name for `errorLike` and then try to run it again with a `new` instance of it.

I have also added new tests to prove this behavior and also tests to cover anonymous functions.

Thanks for your attention guys 😄

### CODE REVIEW

**Comment by meeber** [comment]
Nice catch! LGTM

**Comment by lucasfcosta** [comment]
@meeber Actually I've gotta see why the coverage decrease. But thanks for your input 😄 

It is a little bit strange that it did so. I didn't change a single test, I just added new ones.
The changes on code should not decrease coverage also, since it will just go to the other `getFunctionName` branch when the first one was empty.

I'll investigate this tomorrow, if you guys find any reasons for this please let me know.

**Comment by meeber** [comment]
@lucasfcosta It's because there's one line of code that won't get counted toward coverage until the IE sauce tests are run on merge :D

**Comment by keithamus** [comment]
I had a hunch there was an issue like that! I couldn't find the issue in chai's tracker so I let it lie but next time I think I'll speak up!

LGTM. Great work as always @lucasfcosta

**Comment by lucasfcosta** [comment]
Thanks guys, I've forgotten PRs don't run on SauceLabs.

Build appears to be failing on IE, I already know why and how to fix it. I'll do it as soon as I get home.

**Comment by keithamus** [comment]
It's a bit of a bummer that SauceLabs won't run for PRs... I'd love to be able to find a workable solution.

**Comment by meeber** [comment]
@keithamus For modules that auto-publish, does it wait to see if the commit-to-master build fails before publishing?

**Comment by keithamus** [comment]
Yes. Any commit in master which passes all of the tests (across all travis builds) will then publish to npm.

For the last successful build, you can see the following log lines:

 - [548 - the line which shows `npm t` passed](https://travis-ci.org/chaijs/check-error/jobs/134812491#L548)
 - [555 - travis-after-all waiting for the other jobs to pass](https://travis-ci.org/chaijs/check-error/jobs/134812491#L555)
 - [558 - here's where the semantic release stuff happens](https://travis-ci.org/chaijs/check-error/jobs/134812491#L558)
 - [568 - npm confirmation of the publish step](https://travis-ci.org/chaijs/check-error/jobs/134812491#L568)

Conversely, here is the latest build which is failing: 

 - [574 - the line which shows `npm t` failed](https://travis-ci.org/chaijs/check-error/jobs/135544062#L574)
 - ...nothing of value after that

The way this works is as follows:
 - [L14-15 of .travis.yml tells travis to run `npm t`](https://github.com/chaijs/check-error/blob/master/.travis.yml#L14-L15)
 - [L16-17 tells it travis to run the command on L17 _only if the script command passes_](https://github.com/chaijs/check-error/blob/master/.travis.yml#L16-L17)

To go a little further into the guts, `semantic-release` command will only run if the `TRAVIS_BRANCH` var is `master` (plus some other things, read the full check code [here](https://github.com/semantic-release/condition-travis/blob/next/index.js)) and needs `NPM_TOKEN` (to publish to npm) and `GH_TOKEN` (to generate release notes for the github releases page). These (as well as the `SAUCELABS_ACCESS_KEY` and `SAUCELABS_USERNAME`) are all in [L22-25](https://github.com/chaijs/check-error/blob/master/.travis.yml#L22-L25), encrypted. Travis doesn't decrypt these tokens for PRs because it would be a security breach.

In summary:

 - PRs can never publish to npm, or github, or test on saucelabs - because they're all encrypted values
 - `master` will get these tokens
 - `semantic-release` can only ever run after a successful build because of the travis config
 - `semantic-release` will only release when the conditions are right

**Comment by meeber** [comment]
@keithamus Thanks for the detailed explanation! In particular, the `travis-after-all` module is new to me. Very interesting.
